src/ray/gcs/asio.cc

```diff
@@ -67,13 +68,23 @@ void RedisAsioClient::handle_write(boost::system::error_code error_code) {
 }
 
 void RedisAsioClient::add_read() {
+  // Socket is non-thread safe, dispatch read to backend thread.
+  io_service_.dispatch(boost::bind(&RedisAsioClient::do_add_read, this));
+}
+
+void RedisAsioClient::do_add_read() {
   read_requested_ = true;
   operate();
 }
 
 void RedisAsioClient::del_read() { read_requested_ = false; }
 
 void RedisAsioClient::add_write() {
+  // Socket is non-thread safe, dispatch write to backend thread.
+  io_service_.dispatch(boost::bind(&RedisAsioClient::do_add_write, this));
+}
+
+void RedisAsioClient::do_add_write() {
```

Comment:
Can we just make this a lambda in `add_write` for readability?

Reply:
same for `do_add_read` below

Reply:
Good point.